li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -0,0 +1,342 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class LiKafkaInstrumentedProducerImpl<K, V> implements Producer<K, V>, EventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LiKafkaInstrumentedProducerImpl.class);
+
+  private final ReadWriteLock delegateLock = new ReentrantReadWriteLock();
+  private final Properties baseConfig;
+  private final Map<String, String> translatedBaseConfig;
+  private final Function<Map<String, String>, Producer<K, V>> producerFactory;
+  private final String mdsUrl;
+  private final CountDownLatch initialConnectionLatch = new CountDownLatch(1);
+  private final MetricsProxy metricsProxy = new MetricsProxy();
+  private volatile Long closedAt = null;
+  private volatile Map<String, String> configOverrides;
+  private volatile Producer<K, V> delegate;
+  private SimpleClient mdsClient;
+
+  public LiKafkaInstrumentedProducerImpl(
+      Properties baseConfig,
+      Function<Map<String, String>, Producer<K, V>> producerFactory,
+      String mdsUrl
+  ) {
+    List<String> conversionIssues = new ArrayList<>(1);
+    this.baseConfig = baseConfig;
+    this.translatedBaseConfig = LiKafkaClientsUtils.toStringMap(baseConfig, conversionIssues);
+    this.producerFactory = producerFactory;
+    this.mdsUrl = mdsUrl;
+
+    if (!conversionIssues.isEmpty()) {
+      StringJoiner csv = new StringJoiner(", ");
+      conversionIssues.forEach(csv::add);
+      LOG.error("issues translating producer config to strings: {}", csv);
+    }
+
+    mdsClient = new SimpleClient(
+        mdsUrl,
+        TimeUnit.MINUTES.toMillis(1),
+        TimeUnit.HOURS.toMillis(1),
+        translatedBaseConfig,
+        this
+    );
+
+    boolean tryFallback;
+    Exception issue = null;
+    try {
+      tryFallback = initialConnectionLatch.await(30, TimeUnit.SECONDS);
+    } catch (Exception e) {
+      tryFallback = true;
+      issue = e;
+    }
+
+    if (tryFallback) {
+      boolean delegateChanged = recreateDelegate(true);
+      if (delegateChanged) {
+        if (issue != null) {
```

Comment:
I don't understand this `if-else` block. What is the difference in the conveyed information in both log statements? 

I think it is useful to add a log message in the `catch` block near Line 83, along with the exception thrown.

Reply:
the exception caught is always printed, but either as error or as a warning (in case out timeout actually cared with a successful connection)